base/inference.jl

```diff
@@ -821,7 +810,109 @@ function abstract_apply(af, fargs, aargtypes::Vector{Any}, vtypes, sv, e)
     return abstract_call(af, (), Any[Vararg{Any}], vtypes, sv, ())
 end
 
+function isconstantargs(args, argtypes::Vector{Any}, sv::StaticVarInfo)
+    if isempty(argtypes)
+        return true
+    end
+    if is(args,()) || isa(argtypes[end], Vararg)
+        return false
+    end
+    for i = 1:length(args)
+        arg = args[i]
+        t = argtypes[i]
+        if !isType(t) || has_typevars(t.parameters[1])
+            if isconstantref(arg, sv) === false
+                return false
+            end
+        end
+    end
+    return true
+end
+
+function _ieval_args(args, argtypes::Vector{Any}, sv::StaticVarInfo)
+    c = cell(length(args))
+    for i = 1:length(argtypes)
+        t = argtypes[i]
+        if isType(t) && !has_typevars(t.parameters[1])
+            c[i] = t.parameters[1]
+        else
+            c[i] = _ieval(isconstantref(args[i], sv))
+        end
+    end
+    return c
+end
+
+@pure function type_typeof(v::ANY)
+    if isa(v, Type)
+        return Type{v}
+    end
+    return typeof(v)
+end
+
+function pure_eval_call(f, fargs, argtypes, sv, e)
+    if !isa(f, Function) # TODO: maybe replace with :call?
+        return false
+    end
+    if !isconstantargs(fargs, argtypes, sv)
+        return false
+    end
+
+    args = _ieval_args(fargs, argtypes, sv)
+    tm = _topmod()
+    if isgeneric(f)
+        atype = Tuple{Any[type_typeof(a) for a in args]...}
+        meth = _methods(f, atype, 1)
+        if meth === false || length(meth) != 1
+            return false
+        end
+        meth = meth[1]::SimpleVector
+        linfo = try
+            func_for_method(meth[3], atype, meth[2])
+        catch
+            NF
+        end
+        if linfo === NF
+            return false
+        end
+        if !linfo.pure
+            typeinf(linfo, meth[1], meth[2], linfo)
+            if !linfo.pure
+                return false
+            end
+        end
+    elseif !isdefined(f, :code) || !f.code.pure
+        return false
+    end
+
+    local v
+    try
+        v = f(args...)
```

Comment:
even if we trust the user that the function is pure, this looks dangerous. pure usually does not imply terminating (or reasonably fast anyway) so we should at least not use that name.